### ISSUE TITLE
Message lifetimes

### ISSUE THREAD

**Issue by Biep** [vector-im/element-meta#364]

**Comment by ghost** [comment]
It would be extremely helpful if deleted messages would be overwritten, so that recovery is impossible. I'm not sure if this is already planned or if I'm missing something, but just saying. I think it's crucial to the use case of message lifetimes, as this feature will attract a lot of activists who may not be aware of this risk.

**Comment by KopfKrieg** [comment]
> It would be extremely helpful if deleted messages would be overwritten, so that recovery is impossible.

I just want to point out that overwriting data is on nearly all currently used flash-media useless (due to life-prolonging measures of the media like wear-leveling, and also due to the nature of the used filesystem, which may also implement features like wear-leveling, or copy-on-write, etc.).

**Comment by ghost** [comment]
Oh, thanks for the info! It would be great then if there would be some technique that would make recovery more difficult.

**Comment by Biep** [comment]
And also - one only has command over one's own computer.  Others may ignore any request to delete.  And back-ups and archives will typically not be adapted, so the deletion is inherently unsafe.

**Comment by crti** [comment]
I would love to invite contacts to the Matrix/Element world (from Signal or WA) but features like disappearing messages after a custom timeframe are essential IMHO. Haven't understood much about the difference server and group member apps and a false "sense of security", I think even if working only on the member apps this would be important. Hope in the next months an introduction will be possible, so that Element has a chance to catch up to Signal at least.

**Comment by tmonz** [comment]
I'd be highly in favour of having such a feature added.

**Comment by 3nprob** [comment]
It seems that in this thread there is discussion both about absolute message lifetimes (as per this issue) and self-destructing messages (#82). Just want to note that as described these are not the same feature.

#82 describes supporting [MSC2228](https://github.com/matrix-org/matrix-spec-proposals/pull/2228).

Lifetimes sets an expiration/destruction time at creation. After that time, the message should no longer be available. A given use-case was a "happy birthday" message which should not be sent after the birthday has passed.

@Biep would you say that this feature (lifetimes) is still relevant if #82 (ephemeral messages) is implemented?

**Comment by Biep** [comment]
Yes, very much so.  But I am afraid that implementing a half-solution (only ephemeral messages) will push a thorough solution off the table.  (Not unlike the way partial solutions for spaces and threads, and maybe eventually captions, will no doubt push a generic solution for grouping - that would have enabled implementing all those specific cases - off the table.)

Destruction is merely one specific aspect of message lifetimes.  The whole notion of "keeping on trying to send" (part of the client-side aspect of the lifetimes proposal) is not even addressed with ephemeral messages - but I can tell you that it is vital in countries where connections are erratic.  My experiences is in West-Africa, where connections come and go, and people may write tens of messages while without a connection - and then, currently, need to keep revisit all those rooms to tell Element to try again sending, until the action succeeds.  Others only have connections through public WiFi, such as in supermarkets (no money for mobile Internet), and want their messages to be sent automatically when going there - but not after they have stopped being relevant.
Having a unified model for lifetimes is so much better than a set of patches for some of the more common use cases - again comparable to generic grouping vs. special-cased spaces, threads, captions, and so on.

**Comment by Hexalyse** [comment]
Have we really been discussing this for two years ? I feel like, even just ephemeral messages took ages to be implemented on some clients, even after it was implemented on Synapse. I can't even imagine how something like what you propose @Biep will take to be discussed, and eventually implemented. 

But your proposal indeed sounds like a more generic and powerful solution that would allow many different use cases client-side (but this should be discussed on the Synapse repo or even on the repo for the specs of matrix since it's there that everything begins, not on Element repo which is a client).

In the meantime, I've personally stopped using matrix because of the lack of adoption by my contacts, and the pace at which it evolved compared to less privacy-friendly but way more full-featured IMs like Telegram and the like. I may come back to matrix if all those things end up implemented. I unfortunately don't feel like I have the skill level to dive into matrix server development.

**Comment by 3nprob** [comment]
> not on Element repo which is a client

Indeed. 

I guess the next step would be to draft an MSC and file it in https://github.com/matrix-org/matrix-spec-proposals/pulls, as it would require extending the protocol.

**Comment by Biep** [comment]
That would go beyond my capabilities, I am afraid.
